### ISSUE TITLE
Honor DNS TTL in dynamic upstreams

### ISSUE THREAD

**Issue by hermandavid** [caddyserver/caddy#4853]
Currently, there is a static configuration option that controlls the TTL of dynamic upstreams response. Is there any specific reason why caddy couldn't get the upstream TTL from DNS response?

**Comment by mholt** [comment]
TTL is for DNS resolvers to know how long to cache their lookups. The application does not generally need to know TTL to work effectively. The only reason we have a `Refresh` parameter is because severs can be extremely high-traffic and we want to avoid any unnecessary kernel/network communication. If/when we _do_ call out to the kernel (or DNS directly in a Go implementation of a DNS resolver), then we use TTL to know whether we need an authoritative lookup to get the latest results.

Basically, DNS TTL and Refresh have different purposes.

But that might just be an opinion of mine, I don't know. I'm sure there are arguments both ways.

**Comment by hermandavid** [comment]
I understand your arguments, and I might have misinterpreted the prupose of the TTL value in the DNS response.

My thinking was that it would increase the flexibility and it could make the whole communication more effective if done properly by the DNS server. The cache TTL could be different for each host and DNS server could also handle domains that would otherwise fallback to static domains by responding with lower TTL and thus preventing caddy for trying to resolve dynamic upstream for every single request even if we know beforehand that it would fail.

**Comment by mholt** [comment]
@hermandavid I guess I understand. However, Go does not expose the TTL to us: https://pkg.go.dev/net#SRV

Which is another reason why I suspect the DNS TTL is to benefit the DNS resolvers so that they can cache lookups and allow frequent lookups to be efficient.

Hence, probably in your case, it'd be best to set a low `Refresh` interval so that the DNS lookups happen frequently. Then set your TTL values accordingly to the timespans you desire. Make sense? I don't think we need to implement this a second time in Caddy (probably -- I could be wrong, but some significant performance testing in real world scenarios would need to take place).

Thanks for the feature request; I'm going to close this for now though. Feel free to continue discussion if needed!
